doc/state-tutorial.md

```diff
@@ -0,0 +1,411 @@
+
+// There are no side-effects for get, so nothing to do
+def update(s: State, i: Get): State =
+  s
+```
+
+### Ensure
+
+This is the "post condition". What did we expect to happen in the real world vs
+our model?
+
+Just as a side note, the `s0` means the state before update is run, and `s1` is
+the current state. It's always possible, and fairly common, to just use `s1` and
+the input value.
+
+```scala
+// Almost the reverse of update, for side-effect operations we may not observe anything just yet
+def ensure(s0: State, s1: State, i: Put, o: Unit): Result =
+  Result.success
+
+def ensure(s0: State, s1: State, i: Get, o: Option[String]): Result =
+  s1.map.get(i.key) ==== o
+```
+
+## FAQ
+
+The common question here is "won't we just end up with two implementations"? Yes
+and no. Obviously this is a trivial example. The model is so simple that we can
+actually have a 100% full re-implementation. This won't always be the case, we
+may only want to guarantee some invariants are true, but not the exact
+behaviour. Also remember that we haven't actually looked at the implementation
+of KV yet. It might be storing keys on files, or across the network with
+multiple copies. If there are any bugs in any of that code it's quite likely we
+will find it at some point.
+
+## Sequential
+
+So the final piece is to hook up all our commands to form a property. It would look something like this:
+
+```scala
+def testKVFileSequential: Property = {
+  val root = File("tmpdir")
+  // Create a simple file-system KV store
+  val kv = KV.file(root)
+  sequential(
+    // The range of commands to generate
+      Range.linear(1, 100)
+    // Out initial state to be used for each test run
+    , State(Map())
+    // The list of commands
+    , List(putCommand, getCommand)
+    // Any cleanup code for the real-world, which in this case is to delete the KV file root for each test run
+    , () => root.delete()
+    )
+}
+```
+
+This can be added to Hedgehog like any other property:
+
+```scala
+override def tests: List[Test] =
+  List(
+    property("test file KV (sequential)", testKVFileSequential)
+  )
+```
+
+Some errors that a naive  file implementation found:
+
+#### Invalid file path
+
+```scala
+> Var(Name(0)) = Put(???,b)
+
+> Error thrown running execute: Invalid file path
+```
+
+The first set of line(s) are the series of commands that have been run in order
+(ignore `Var` for now). Everything after that is our `Result` log.
+
+So I'm generating invalid keys. So there is some restrictions on what our keys
+can be. Interesting. For now update the generator to something much more
+restrictive.
+
+Note that this would have been found pretty quickly by our "normal" put/get
+property. So far state-based testing hasn't done much for us.
+
+Let's fix the key generator.
+
+```scala
+def genKey : Gen[String] =
+  for {
+    c <- Gen.lower
+    s <- Gen.string(Gen.choice1(Gen.lower, Gen.constant('/')), Range.linear(0, 10))
+  } yield c + s
```

Comment:
As you can see, build for Scala 2.13.0 failed because of this. 
The error message suggests using String interpolation but maybe just `c +: s`?